File: HDF4Examples/FORTRAN/VD/h4ex_VD_set_get_vdata_attr.f
      program vdata_attributes 
      implicit none
C
C     Parameter declaration
C
      character*18 FILE_NAME
      character*14 VDATA_NAME
      character*4  FIELD_NAME
      character*10 VATTR_NAME
      character*6  FATTR_NAME
      integer      VATTR_N_VALUES, FATTR_N_VALUES 
C
      parameter (FILE_NAME    = 'General_Vdatas.hdf',
     +           VDATA_NAME   = 'Solid Particle',
     +           FIELD_NAME   = 'Mass',
     +           VATTR_NAME   = 'Site Ident',
     +           FATTR_NAME   = 'Scales')
      parameter (VATTR_N_VALUES = 3,
     +           FATTR_N_VALUES = 4)
   
      integer DFACC_WRITE, FULL_INTERLACE, HDF_VDATA
      integer DFNT_INT32, DFNT_CHAR8
      parameter (DFACC_WRITE    =  2,
     +           FULL_INTERLACE =  0,
     +           HDF_VDATA      = -1,
     +           DFNT_INT32     = 24,
     +           DFNT_CHAR8     =  4)
C
C     Function declaration
C
      integer hopen, hclose
      integer vfstart, vsffnd, vsfatch, vsfscat, vsfsnat, 
     +        vsffnas, vsffidx, vsfnats, vsfainf, vsfgcat, vsfgnat,
     +        vsfdtch, vfend 

C
C**** Variable declaration *******************************************
C
      integer   status
      integer   file_id, vdata_id, vdata_ref
      integer   field_index, n_vdattrs, n_fldattrs
      integer   vdata_type, vdata_n_values, vdata_size
      integer   field_type, field_n_values, field_size 
      character vd_attr(VATTR_N_VALUES)
      integer   fld_attr(FATTR_N_VALUES)
      character vattr_buf(VATTR_N_VALUES)
      integer   fattr_buf(FATTR_N_VALUES)
      character vattr_name_out(30), fattr_name_out(30)
      data vd_attr /'A', 'B', 'C'/
      data fld_attr /2, 4, 6, 8/
C
C**** End of variable declaration ************************************
C
C
C     Open the HDF file for writing.
C
      file_id = hopen(FILE_NAME, DFACC_WRITE, 0)
C
C     Initialize the VS interface.
C
      status = vfstart(file_id) 
C
C     Get the reference number of the vdata named VDATA_NAME.
C     
      vdata_ref = vsffnd(file_id, VDATA_NAME)
C
C     Attach to the vdata for writing. 
C
      vdata_id = vsfatch(file_id, vdata_ref, 'w') 
C
C     Attach an attribute to the vdata, as it is indicated by second parameter.
C
      status = vsfscat(vdata_id, HDF_VDATA, VATTR_NAME, DFNT_CHAR8,
     +                  VATTR_N_VALUES, vd_attr)
C
C     Get the index of the field FIELD_NAME within the vdata.
C
      status = vsffidx(vdata_id, FIELD_NAME, field_index)
C
C     Attach an attribute to the field with the index field_index.
C
      status = vsfsnat(vdata_id, field_index, FATTR_NAME, DFNT_INT32,
     +                  FATTR_N_VALUES, fld_attr)
      
C
C     Get the number of attributes attached to the vdata's first
C     field - should be 0.
C
      n_fldattrs = vsffnas(vdata_id, 0)
      write(*,*) 'Number of attributes of the first field' 
      write(*,*) ' of the vdata: ', n_fldattrs
C
C     Get the number of the attributes attached to the field specified by
C     index field_index - should be 1.
C
      n_fldattrs = vsffnas(vdata_id, field_index)
      write(*,*) 'Number of attributes of field ', FIELD_NAME,
     +           n_fldattrs
C
C     Get the total number of the field's and vdata's attributes - should be 2.
C
      n_vdattrs = vsfnats(vdata_id)
      write(*,*) 'Number of attributes of the vdata and its fields: ',
     +           n_vdattrs
C
C     Get information about the vdata's first attribute, indicated by 
C     the third parameter, which is the index of the attribute.
C
      status = vsfainf(vdata_id, HDF_VDATA, 0, vattr_name_out,
     +                 vdata_type, vdata_n_values, vdata_size)
C
C     Get information about the first attribute of the field specified by
C     field_index.
C
      status = vsfainf(vdata_id, field_index, 0, fattr_name_out, 
     +                 field_type, field_n_values, field_size)
C
C     Get the vdata's first attribute.
C
      status = vsfgcat(vdata_id, HDF_VDATA, 0, vattr_buf)
      write(*,*) 'Values of vdata attribute  ', vattr_buf
C
C     Get the first attribute of the field specified by field_index.
C
      status = vsfgnat(vdata_id, field_index, 0, fattr_buf)
      write(*,*)  'Values of the field attribute = ', fattr_buf
C       
C     Terminate access to the vdata and to the VS interface, and
C     close the HDF file.
C
      status = vsfdtch(vdata_id)
      status = vfend(file_id)
      status = hclose(file_id)
      end

File: HDF4Examples/C/VD/h4ex_VD_set_get_vdata_attr.c
#include <stdio.h>

#include "hdf.h"

#define FILE_NAME      "General_Vdatas2.hdf"
#define VDATA_NAME     "Solid Particle"
#define FIELD_NAME     "Mass"
#define VATTR_NAME     "Site Ident" /* name of the vdata attribute  */
#define FATTR_NAME     "Scales"     /* name of the field attribute  */
#define VATTR_N_VALUES 3            /* number of values in the vdata attribute */
#define FATTR_N_VALUES 4            /* number of values in the field attribute */

int
main()
{
    /************************* Variable declaration **************************/

    int32 file_id, vdata_ref, vdata_id;
    int32 field_index,                                /* index of a field within the vdata */
        n_vdattrs,                                    /* number of vdata attributes */
        n_fldattrs,                                   /* number of field attributes */
        vdata_type,                                   /* to hold the type of vdata's attribute */
        vdata_n_values,                               /* to hold the number of vdata's attribute values   */
        vdata_size,                                   /* to hold the size of vdata's attribute values     */
        field_type,                                   /* to hold the type of field's attribute            */
        field_n_values,                               /* to hold the number of field's attribute values   */
        field_size;                                   /* to hold the size of field's attribute values     */
    char  vd_attr[VATTR_N_VALUES]  = {'A', 'B', 'C'}; /* vdata attribute values*/
    int32 fld_attr[FATTR_N_VALUES] = {2, 4, 6, 8};    /* field attribute values*/
    char  vattr_buf[VATTR_N_VALUES];                  /* to hold vdata attribute's values */
    int32 fattr_buf[FATTR_N_VALUES];                  /* to hold field attribute's values */
    char  vattr_name[30],                             /* name of vdata attribute */
        fattr_name[30];                               /* name of field attribute */

    /********************** End of variable declaration **********************/

    /*
     * Open the HDF file for writing.
     */
    if ((file_id = Hopen(FILE_NAME, DFACC_WRITE, 0)) == FAIL)
        printf("*** ERROR from Hopen\n");

    /*
     * Initialize the VS interface.
     */
    if (Vstart(file_id) == FAIL)
        printf("*** ERROR from Vstart\n");

    /*
     * Get the reference number of the vdata named VDATA_NAME.
     */
    vdata_ref = VSfind(file_id, VDATA_NAME);

    /*
     * Attach to the vdata for writing.
     */
    vdata_id = VSattach(file_id, vdata_ref, "w");

    /*
     * Attach an attribute to the vdata, i.e., indicated by the second parameter.
     */
    if (VSsetattr(vdata_id, _HDF_VDATA, VATTR_NAME, DFNT_CHAR, VATTR_N_VALUES, vd_attr) == FAIL)
        printf("*** ERROR from VSsetattr\n");

    /*
     * Get the index of the field FIELD_NAME within the vdata.
     */
    if (VSfindex(vdata_id, FIELD_NAME, &field_index) == FAIL)
        printf("*** ERROR from VSfindex\n");

    /*
     * Attach an attribute to the field field_index.
     */
    if (VSsetattr(vdata_id, field_index, FATTR_NAME, DFNT_INT32, FATTR_N_VALUES, fld_attr) == FAIL)
        printf("*** ERROR from VSsetattr\n");

    /*
     * Get the number of attributes attached to the vdata's first
     * field - should be 0.
     */
    n_fldattrs = VSfnattrs(vdata_id, 0);
    printf("Number of attributes of the first field of the vdata: %d\n", n_fldattrs);

    /*
     * Get the number of attributes attached to the field specified by
     * field_index - should be 1.
     */
    n_fldattrs = VSfnattrs(vdata_id, field_index);
    printf("Number of attributes of field %s: %d\n", FIELD_NAME, n_fldattrs);

    /*
     * Get the total number of the field's and vdata's attributes - should be 2.
     */
    n_vdattrs = VSnattrs(vdata_id);
    printf("Number of attributes of the vdata and its fields: %d\n", n_vdattrs);

    /*
     * Get information about the vdata's first attribute, indicated
     * by the third parameter which is the index of the attribute.
     */
    if (VSattrinfo(vdata_id, _HDF_VDATA, 0, vattr_name, &vdata_type, &vdata_n_values, &vdata_size) == FAIL)
        printf("*** ERROR from VSattrinfo\n");

    /*
     * Get information about the first attribute of the field specified by
     * field_index.
     */
    if (VSattrinfo(vdata_id, field_index, 0, fattr_name, &field_type, &field_n_values, &field_size) == FAIL)
        printf("*** ERROR from VSattrinfo\n");

    /*
     * Get the vdata's first attribute.
     */
    if (VSgetattr(vdata_id, _HDF_VDATA, 0, vattr_buf) == FAIL)
        printf("*** ERROR from VSgetattr\n");
    printf("Values of the vdata attribute = %c %c %c\n", vattr_buf[0], vattr_buf[1], vattr_buf[2]);

    /*
     * Get the first attribute of the field specified by field_index.
     */
    if (VSgetattr(vdata_id, field_index, 0, fattr_buf) == FAIL)
        printf("*** ERROR from VSgetattr\n");
    printf("Values of the field attribute = %d %d %d %d\n", fattr_buf[0], fattr_buf[1], fattr_buf[2],
           fattr_buf[3]);

    /*
     * Terminate access to the vdata and to the VS interface, then close
     * the HDF file.
     */
    if (VSdetach(vdata_id) == FAIL)
        printf("*** ERROR from VSdetach\n");
    if (Vend(file_id) == FAIL)
        printf("*** ERROR from Vend\n");
    if (Hclose(file_id) == FAIL)
        printf("*** ERROR from Hclose\n");

    return 0;
}
